ml-proto/test/int32.wasm

```diff
@@ -0,0 +1,37 @@
+(; Int arith operations ;)
+
+(module
+  (func $clz32 (param $x i32) (result i32)
+    (i32.clz (get_local $x))
+  )
+
+  (func $ctz32 (param $x i32) (result i32)
+    (i32.ctz (get_local $x))
+  )
+
+  (func $pcnt32 (param $x i32) (result i32)
+    (i32.popcnt (get_local $x))
+  )
```

Comment:
I may be wrong, but IIRC, the “population count” terminology long precedes Intel. It is used in GHC base package, GCC, Sparc, Cray, and CDC (the latter ones long before Intel). Even the Mark I had it but called it Sideways Add.

From: rossberg-chromium [mailto:notifications@github.com]
Sent: Wednesday, September 16, 2015 12:33 AM
To: WebAssembly/spec
Cc: Kuan, George
Subject: Re: [spec] implementation and tests for clz, ctz, and popcnt (#50)


In ml-proto/test/int32.wasm<https://github.com/WebAssembly/spec/pull/50#discussion_r39600610>:

> @@ -0,0 +1,37 @@

> +(; Int arith operations ;)

> +

> +(module

> +  (func $clz32 (param $x i32) (result i32)

> +    (i32.clz (get_local $x))

> +  )

> +

> +  (func $ctz32 (param $x i32) (result i32)

> +    (i32.ctz (get_local $x))

> +  )

> +

> +  (func $pcnt32 (param $x i32) (result i32)

> +    (i32.popcnt (get_local $x))

> +  )
On 16 September 2015 at 01:36, Dan Gohman <notifications@github.com> wrote: In ml-proto/test/int32.wasm <https://github.com/WebAssembly/spec/pull/50#discussion_r39579393>: > @@ -0,0 +1,37 @@ > +(; Int arith operations ;) > + > +(module > + (func $clz32 (param $x i32) (result i32) > + (i32.clz (get_local $x)) > + ) > + > + (func $ctz32 (param $x i32) (result i32) > + (i32.ctz (get_local $x)) > + ) > + > + (func $pcnt32 (param $x i32) (result i32) > + (i32.popcnt (get_local $x)) > + ) It'd be nice to keep the function name identical to the opcode name.
Btw, despite it being what Intel picked on a bad day, `popcnt` really is an odd name out. It is both inconsistent with the other count operators, and it sounds like some kind of stack operator. Would anybody mind if we changed it? For example, how about `cnz` (count non-zeros), which would be in line with the other count operators. (If you say it's a worthwhile idea I can create an issue or PR.)

—
Reply to this email directly or view it on GitHub<https://github.com/WebAssembly/spec/pull/50/files#r39600610>.